design/one-pager-change-logs.md

```diff
@@ -0,0 +1,310 @@
+and export the change logs is to use the [OpenTelemetry
+Collector](https://opentelemetry.io/docs/collector/), but a more details
+analysis of that setup is out of scope for this document.
+
+#### Pod Log Benefits
+
+Pod logs are a good option for the change log location because they:
+
+* are very standardized, with many tools capable of reading, interacting with,
+  and scraping them
+* have built in management features like log rotation
+* provide sufficient durability for our needs, e.g. they can survive network
+  outages and pod crashes
+
+### Flow of Change Log Entries
+
+As mentioned in the previous section, change log entries will be written to the
+pod logs. However, there is some additional structure we will build in order to
+keep change logs entirely separated from the regular provider log entries, which
+will add not only with human readability, but also post processing and analysis
+by the system responsible for long term storage.
+
+Each provider pod that generates change log entries will have a sidecar
+container solely responsible for receiving the entries and writing them to
+`stdout` (i.e. the pod logs for the sidecar container). This sidecar container
+will be listening on a gRPC connection from the main provider container. This
+gRPC connection between the two containers will be established over a Unix
+domain socket on a shared `emptyDir` volume mounted by both containers. This
+setup allows the main provider container to send the change log entries to the
+sidecar container in a fairly performant manner without needed any network
+access at all.
+
+The change log entry data will be serialized in a binary protobuf format for
+transmission over the gRPC connection, while the data will be persisted to the
+pod logs in a JSON serialized format.
+
+#### Summary: Change Log Entry Flow
+
+* The provider will have sidecar container where each entry is sent
+* Main container will communicate through gRPC with the sidecar container over a
+  Unix domain socket on a shared volume
+* Data is serialized in a protobuf binary format and written to sidecar
+  container logs in JSON format
+
+### General Architecture
+
+The general architecture of the change log system is captured in the diagram
+below:
+
+![Change logs architecture diagram](./images/change-logs-architecture.png)
+
+### Implementation Rollout
+
+This feature will be declared at an
+[alpha](https://docs.crossplane.io/latest/learn/feature-lifecycle/#alpha-features)
+level of maturity when it is first released, meaning that Crossplane users must
+opt-in to using this functionality by setting an `--enable-change-logs` feature
+flag. Note this will flag will need to be set on each provider for which change
+logs are desired, likely via a `DeploymentRuntimeConfig`.
+
+Also note in the plan below that core Crossplane itself does not need to updated
+at all to support this feature. The key changes are in `crossplane-runtime`, the
+Upjet framework, and the ecosystem of providers.
```

Comment:
Thank you very much for pointing this out @sttts! You opened my eyes to the fact that I made a naive assumption about how the upjet providers work and in general how a provider can influence its pod spec.  spoiler alert: it can't 😂 

I had assumed that at build time a provider specifies details about its pod spec, including its main container and any init or sidecar containers it wants as well. This is not true. The package manager is what crafts the provider's deployment/pod spec.  I had forgotten this 😊 

So I will definitely be looking into this in more detail, but some early thoughts:

* I'll see what's possible with a `DeploymentRuntimeConfig`. Perhaps a sidecar container snippet can be specified in there and then it gets merged nicely with other config the package manager will enforce 🙏 
* If not, the package manager could be updated to inject a sidecar container for each provider where change logs are desired. This would indeed require a code change to core crossplane, so thanks again for pointing this gap out @sttts!! 🙇‍♂️

Reply:
I was really pleased to see how well a `DeploymentRuntimeConfig` could be used to set this architecture up for a provider during my prototyping efforts last week. I've made the recommendation that at least for this alpha implementation, we can just use a `DeploymentRuntimeConfig` instead of having to make any changes to core crossplane and/or its package manager. I've also included an example of what this `DeploymentRuntimeConfig` looks like.